File: typespec/admin/users/invitations.go
// Package users contains Admin API user and invitation wire types.
package users

import (
	"time"

	adminspec "github.com/vetchium/src/typespec/admin"
	"github.com/vetchium/src/typespec/admin/authorization"
	"github.com/vetchium/src/typespec/common"
)

type AdminInvitationID string
type AdminInvitationToken common.OpaqueToken

type InviteUserRequest struct {
	EmailAddress common.EmailAddress               `json:"email_address"`
	Permissions  []authorization.AdminPermissionID `json:"permissions,omitempty"`
}

func (r InviteUserRequest) Normalize() InviteUserRequest {
	r.EmailAddress = common.NormalizeEmailAddress(r.EmailAddress)
	return r
}

func (r InviteUserRequest) Validate() []string {
	fields := make([]string, 0, 2)
	if !common.IsEmailAddress(r.Normalize().EmailAddress) {
		fields = append(fields, "email_address")
	}
	if !authorization.ValidatePermissions(r.Permissions) {
		fields = append(fields, "permissions")
	}
	return fields
}

type InviteUserResponse struct {
	AdminInvitationID AdminInvitationID `json:"admin_invitation_id"`
	ExpiresAt         time.Time         `json:"expires_at"`
}

type CompleteSetupRequest struct {
	InvitationToken   AdminInvitationToken  `json:"invitation_token"`
	Password          common.NewPassword    `json:"password"`
	DisplayName       common.DisplayName    `json:"display_name"`
	PreferredLanguage common.FrontendLocale `json:"preferred_language"`
}

func (r CompleteSetupRequest) Normalize() CompleteSetupRequest {
	r.DisplayName = common.NormalizeDisplayName(r.DisplayName)
	return r
}

func (r CompleteSetupRequest) Validate() []string {
	r = r.Normalize()
	fields := make([]string, 0, 4)
	if !common.IsOpaqueToken(string(r.InvitationToken)) {
		fields = append(fields, "invitation_token")
	}
	if !common.IsNewPassword(r.Password) {
		fields = append(fields, "password")
	}
	if !common.IsDisplayName(r.DisplayName) {
		fields = append(fields, "display_name")
	}
	if !common.IsFrontendLocale(r.PreferredLanguage) {
		fields = append(fields, "preferred_language")
	}
	return fields
}

type CompleteSetupResponse struct {
	AdminUserID adminspec.AdminUserID `json:"admin_user_id"`
}

func IsAdminInvitationID(value AdminInvitationID) bool {
	return adminspec.IsAdminUserID(adminspec.AdminUserID(value))
}

File: typespec/admin/users/invitations.ts
import type { NewPassword, OpaqueToken } from "../../common/authentication.ts";
import { isNewPassword, isOpaqueToken } from "../../common/authentication.ts";
import type { EmailAddress } from "../../common/common.ts";
import { isEmailAddress, normalizeEmailAddress } from "../../common/common.ts";
import type { DisplayName, FrontendLocale } from "../../common/localization.ts";
import {
  isDisplayName,
  isFrontendLocale,
  normalizeDisplayName,
} from "../../common/localization.ts";
import {
  type AdminPermissionID,
  validatePermissions,
} from "../authorization/types.ts";
import { type AdminUserID, isAdminUserID } from "../types.ts";

export type AdminInvitationID = string;
export type AdminInvitationToken = OpaqueToken;

export function isAdminInvitationID(value: AdminInvitationID): boolean {
  return isAdminUserID(value);
}

export interface InviteUserRequest {
  email_address: EmailAddress;
  permissions?: AdminPermissionID[];
}

export function normalizeInviteUserRequest(
  request: InviteUserRequest,
): InviteUserRequest {
  return {
    ...request,
    email_address: normalizeEmailAddress(request.email_address),
  };
}

export function validateInviteUserRequest(
  request: InviteUserRequest,
): string[] {
  const fields: string[] = [];
  if (!isEmailAddress(normalizeInviteUserRequest(request).email_address)) {
    fields.push("email_address");
  }
  if (
    request.permissions !== undefined &&
    !validatePermissions(request.permissions)
  ) {
    fields.push("permissions");
  }
  return fields;
}

export interface InviteUserResponse {
  admin_invitation_id: AdminInvitationID;
  expires_at: string;
}

export interface CompleteSetupRequest {
  invitation_token: AdminInvitationToken;
  password: NewPassword;
  display_name: DisplayName;
  preferred_language: FrontendLocale;
}

export function normalizeCompleteSetupRequest(
  request: CompleteSetupRequest,
): CompleteSetupRequest {
  return {
    ...request,
    display_name: normalizeDisplayName(request.display_name),
  };
}

export function validateCompleteSetupRequest(
  request: CompleteSetupRequest,
): string[] {
  const normalized = normalizeCompleteSetupRequest(request);
  const fields: string[] = [];
  if (!isOpaqueToken(normalized.invitation_token)) {
    fields.push("invitation_token");
  }
  if (!isNewPassword(normalized.password)) {
    fields.push("password");
  }
  if (!isDisplayName(normalized.display_name)) {
    fields.push("display_name");
  }
  if (!isFrontendLocale(normalized.preferred_language)) {
    fields.push("preferred_language");
  }
  return fields;
}

export interface CompleteSetupResponse {
  admin_user_id: AdminUserID;
}
